py/circuitpy_mpconfig.h

```diff
@@ -276,6 +276,13 @@ typedef long mp_off_t;
 #define BOARD_UART_ROOT_POINTER
 #endif
 
+// Support for internal pullups for boards that have them: e.g. nrf52833
+#if CIRCUITPY_I2C_INTERNAL_PULLUPS
+#define I2C_USE_INTERNAL_PULLUPS = true
+#else
+#define I2C_USE_INTERNAL_PULLUPS = false
+#endif
+
```

Comment:
Don't do this; we will only enable internal pullups on the boards that you want to use them on, which will be your own company's board, not any of the existing boards.

Reply:
Ok, you're saying leave this out of `git add`?

Reply:
The idea of this PR is to be able to specify internal pullups at run-time, instead of compile time. Conceivably you would want `board.I2C()` or similar to use internal pullups, but that would be for a specialized board (e.g. your company board), not for any current board. That would be a reason to use a compile-time setting, and the reason I suggested `BOARD_I2C_USE_INTERNAL_PULLUPS`

Reply:
I hear you. 

The only reason I've added a constant at all is to accommodate `common_hal_board_create_i2c(void)`: It calls`common_hal_busio_i2c_construct` which will include the `internal_pullups` param. I figured the least invasive change was to declare a constant rather than change yet another method signature.

Thoughts?

Reply:
Right, I'm just saying check for the constant `CIRCUITPY_BOARD_I2C_USE_INTERNAL_PULLUPS` in `common_hal_board_create_i2c(void)`. But define that constant directly, and then I think you can eliminate `CIRCUITPY_I2C_INTERNAL_PULLUPS` and `I2C_USE_INTERNAL_PULLUPS`.

Reply:
In other words get rid of the older existing current compile-time checks for internal pullups; we are going to do it at runtime instead.

Reply:
Not to be thick-headed, but confused about what you mean by "define that constant directly". 

To clarify, I assume you mean something like

```
mp_obj_t common_hal_board_create_i2c(void) {
    // All callers have either already verified this or come so early that it can't be otherwise.
    assert(i2c_singleton == NULL || common_hal_busio_i2c_deinited(i2c_singleton));
    busio_i2c_obj_t *self = &i2c_obj;
    self->base.type = &busio_i2c_type;

    // ---> Check if board supports internal pullups
    //         and define the constant. e.g.
    if (internal_pullups_are_supported)
          const bool I2C_USE_INTERNAL_PULLUPS = true;
    else
          const bool I2C_USE_INTERNAL_PULLUPS = false;
    

    common_hal_busio_i2c_construct(self, DEFAULT_I2C_BUS_SCL, DEFAULT_I2C_BUS_SDA, I2C_USE_INTERNAL_PULLUPS, 100000, 255);
    i2c_singleton = (mp_obj_t)self;
    return i2c_singleton;
}
```

My question is how can I know at runtime if the board has support for internal pullups? 

This isn't an issue if it's just for my own board, but if it's for the community at large (which is what I'm assuming) it would need to be more robust - I'd need to know if the board supports this.

Reply:
> My question is how can I know at runtime if the board has support for internal pullups?

With the addition of the `internal_pullups` argument to `busio.I2C()`, _any_ board can now support internal pullups. You just have to ask for them by setting `internal_pullups` to be true in the constructor.

The question now for a particular board is whether `board.I2C()` asks for internal pullups _by default_. So I'm sayin to remove the older compile time constants, and add a `CIRCUITPY_BOARD_I2C_USE_INTERNAL_PULLUPS`, which defaults to 0.

Then in the code above, you would just do
```c
mp_obj_t common_hal_board_create_i2c(void) {
  ...
    common_hal_busio_i2c_construct(self, DEFAULT_I2C_BUS_SCL, DEFAULT_I2C_BUS_SDA, CIRCUITPY_BOARD_I2C_USE_INTERNAL_PULLUPS, 100000, 255);
   ...
}
```

Reply:
Ah, I think I understand where I was confused; you're trusting the developer to specify (and therefore, know) whether the board has support for internal pullups - I was going the other way, requiring CPY developers to define this in `.mk` files.

If I understand, you're saying "assume the board doesn't, but allow the developer to tell us it does". If that's true, then I did need to remember `internal_pullups`, and can get this from the `self` pointer: 

```
busio_i2c_obj_t *self = &i2c_obj;

common_hal_busio_i2c_construct(self, DEFAULT_I2C_BUS_SCL, DEFAULT_I2C_BUS_SDA, self->internal_pullups, 100000, 255);
```
Is this more what you had in mind?

Thanks for all your patience, @dhalbert.

Reply:
Did you post and then delete (retract) a comment there? To answer that, nearly all chips have always had the capability of supporting internal pullups, we just didn't expose that. Now we are. It's a choice of the user about whether they are going to connect some I2C devices without pullups. That could happen on any board.

And, there might be some boards, with board-mounted I2C devices, without external pullups. In that case, we _might_ want `board.I2C()` to use internal pullups by default. It might actually be better not to do that, since it hides whether internal pullups are used or not. On code written for your board, you might just go ahead and set `internal_pullups=True` explicitly when creating a `busio.I2C()` for use with the board-mounted I2C cervices. Then we don't need any compile-time constants at all.

There may be some chips that don't have pull-ups; I don't know. For those chips, the common-hal constructor would raise `NotImplementedError` if `internal_pullups=True`, to indicate it was not possible.

Reply:
Ok. I'm confused about some of this, mainly because I'm trying to relate now to set your suggested `CIRCUITPY_BOARD_I2C_USE_INTERNAL_PULLUPS` at runtime, but perhaps it's better to put a pin in this? 

In the interim, I can simply pass `true` to that part of the method and move on to the bigger issue: `i2c.scan()` is not returning any devices, even after I set the pulls up and read a proper voltage (3.3V) on them at runtime.

Reply:
> Ok. I'm confused about some of this, mainly because I'm trying to relate now to set your suggested `CIRCUITPY_BOARD_I2C_USE_INTERNAL_PULLUPS` at runtime, but perhaps it's better to put a pin in this?

I'm suggesting that the compile-time constant need not exist at all. The user code will decide whether or not to use internal pullups.

>In the interim, I can simply pass true to that part of the method and move on to the bigger issue: i2c.scan() is not returning any devices, even after I set the pulls up and read a proper voltage (3.3V) on them at runtime. At this point, I need to figure out what's going wrong there.

I'll build this and try it out.

Are you trying this with a PCA10100 and an external sensor without pullups, or are you trying it on your own board?